File: packages/shared/src/payback.ts
import type { PaybackInputs, PaybackResult } from "./types";

const round = (value: number) => Number(value.toFixed(2));
const months = (target: number, monthlyPayout: number) =>
  monthlyPayout > 0 ? target / monthlyPayout : Number.POSITIVE_INFINITY;

export function calculatePayback(input: PaybackInputs): PaybackResult {
  const targetMultiple = input.targetMultiple ?? 1.5;
  const notCurrentlyRecovering = input.monthlyPayout <= 0;
  const principalMonths = notCurrentlyRecovering ? Number.POSITIVE_INFINITY : months(input.investment, input.monthlyPayout);
  const targetMonths = notCurrentlyRecovering
    ? Number.POSITIVE_INFINITY
    : months(input.investment * targetMultiple, input.monthlyPayout);

  return {
    principalMonths: notCurrentlyRecovering ? Number.POSITIVE_INFINITY : round(principalMonths),
    targetMonths: notCurrentlyRecovering ? Number.POSITIVE_INFINITY : round(targetMonths),
    yearsToPrincipal: notCurrentlyRecovering ? Number.POSITIVE_INFINITY : round(principalMonths / 12),
    yearsToTarget: notCurrentlyRecovering ? Number.POSITIVE_INFINITY : round(targetMonths / 12),
    notCurrentlyRecovering,
  };
}

export function calculateRecoveryPercent(cumulativePayout: number, investment: number) {
  return investment > 0 ? round(Math.min(1, cumulativePayout / investment)) : 0;
}

File: backend/app/services/payback.py
def round2(value: float) -> float:
    return round(value + 1e-12, 2)


def calculate_payback(input: dict) -> dict:
    monthly_payout = float(input["monthlyPayout"])
    target_multiple = float(input.get("targetMultiple", 1.5))
    if monthly_payout <= 0:
        return {
            "principalMonths": float("inf"),
            "targetMonths": float("inf"),
            "yearsToPrincipal": float("inf"),
            "yearsToTarget": float("inf"),
            "notCurrentlyRecovering": True,
        }
    investment = float(input["investment"])
    principal_months = investment / monthly_payout
    target_months = (investment * target_multiple) / monthly_payout
    return {
        "principalMonths": round2(principal_months),
        "targetMonths": round2(target_months),
        "yearsToPrincipal": round2(principal_months / 12),
        "yearsToTarget": round2(target_months / 12),
        "notCurrentlyRecovering": False,
    }
